contracts/plugins/assets/morpho-aave/MorphoAaveV2TokenisedDeposit.sol

```diff
@@ -0,0 +1,118 @@
+// SPDX-License-Identifier: BlueOak-1.0.0
+pragma solidity 0.8.19;
+
+import { ERC20 } from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { ERC20Burnable } from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
+import { IERC20, IERC20Metadata } from "@openzeppelin/contracts/interfaces/IERC20Metadata.sol";
+import { FixLib, _safeWrap, shiftl_toFix, FIX_ONE } from "../../../libraries/Fixed.sol";
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { IMorphoTokenisedDeposit, IMorpho, IRewardsDistributor, IUsersLens } from "./IMorpho.sol";
+import { RewardableERC4626Vault } from "../erc20/RewardableERC4626Vault.sol";
+
+struct MorphoAaveV2TokenisedDepositConfig {
+    IMorpho morphoController;
+    IUsersLens morphoLens;
+    IRewardsDistributor rewardsDistributor;
+    IERC20Metadata underlyingERC20;
+    IERC20Metadata poolToken;
+    ERC20 rewardToken;
+}
+
+contract MorphoAaveV2TokenisedDeposit is IMorphoTokenisedDeposit {
+    IRewardsDistributor public immutable rewardsDistributor;
+    IMorpho public immutable morphoController;
+    IUsersLens public immutable morphoLens;
+    address public immutable poolToken;
+    address public immutable underlying;
+    uint256 private immutable underlyingOne;
+
+    constructor(MorphoAaveV2TokenisedDepositConfig memory config)
+        RewardableERC4626Vault(
+            config.underlyingERC20,
+            string.concat("Tokenised Morpho Position - ", config.poolToken.name()),
+            string.concat("mrp-", config.poolToken.symbol()),
+            config.rewardToken
+        )
+    {
+        underlying = address(config.underlyingERC20);
+        morphoController = config.morphoController;
+        morphoLens = config.morphoLens;
+        poolToken = address(config.poolToken);
+        rewardsDistributor = config.rewardsDistributor;
+        underlyingOne = 10**ERC20(underlying).decimals();
+    }
+
+    // Offchain keepers should periodically call this such that we accumuate rewards
+    function claimOffchainRewards(uint256 _claimable, bytes32[] calldata _proof) external {
+        uint256 delta;
+        uint256 _totalSupply = totalSupply();
+        if (_totalSupply > 0) {
+            uint256 initialBal = rewardToken.balanceOf(address(this));
+            rewardsDistributor.claim(address(this), _claimable, _proof);
```

Comment:
it looks like anyone can claim rewards for anyone else, which means someone could claim for this wrapper, locking the tokens in it.

if we use a storage variable instead of `initialBal`, we can reference the last reward balance in this calculation.  we would also want to utilize the `_claimAssetRewards()` function to update `rewardsPerShare`.

Reply:
Isn't the idea that anyone can call this function, passing in the wrapper's address? Then in lines 59/60 they get their share of rewards? Maybe I'm misunderstanding something...

Reply:
The only args are the amount to claim and the proof

Reply:
(there could still be something wrong with the reward math I'm not seeing)

Reply:
@pmckelvy1 

@tbrent's assumption is correct. A keeper bot would have to generate a proof for the `MorphoAaveV2TokenisedDeposit` once in a while, and claim the rewards on behalf of the vault.

In this model, only the claimed Morpho tokens (via the `claimOffchainRewards`) method would be distributed to token holders.

Reply:
@pmckelvy1 I've added a test to show I intended Rewards to be claimed. Currently however rewards can actually claimed out of the Token vault, so they will just accumulate over time.

This is Morpho tokens are currently not transferable. Unless your address is whitelisted you're not allowed to transfer held tokens.

If we get Morpho token vault with a siginificant position we could ask the Morpho team nicely if they would be willing to whitelist the vault so users can claim their accumulated reward.

Reply:
@pmckelvy1 @tbrent I spoke with Patrick again. His concern is;

Since any can claim rewards for anyone in morpho. All they need to do is submit the proof and address. It's possible for someone to claim rewards that end up unaccounted.

I think this is a really good point, and is actually points to a more generic case that the`RewardableERC20` does not handle. What happens if reward tokens are just transfered to the token vault outside of us claiming them? These tokens would currently not be accounted for. On low fee chains, it's is not that unusual to see direct airdrops from protocols, and with Morpho, you don't have to be the one initiating a claim to receive rewards.

I'll make a small modification to the `RewardableERC20` that should handle it. This actually simplifies the Morpho implementation a lot.

Reply:
Ah, @pmckelvy1 good catch. Yes. Need storage var.